feat(dotnet11): add process-api-net11 skill for new process APIs (#866)
dotnet/skills: plugins/dotnet11/skills/process-api-net11/SKILL.md

---
name: process-api-net11
description: >
  Provides guidance on the new System.Diagnostics.Process APIs introduced in .NET 11.
  It covers high-level convenience methods (Process.Run, Process.RunAndCaptureText, Process.StartAndForget),
  reliable deadlock-free output reading (Process.ReadAllText/Bytes/Lines), and lifecycle/handle management
  (KillOnParentExit, InheritedHandles, StartDetached).
  Use when starting, orchestrating, or capturing output from external processes in .NET 11 applications.
license: MIT
---

# Process API Improvements — .NET 11

New APIs added to `System.Diagnostics.Process` in .NET 11 simplify process management, eliminate boilerplate, and prevent common deadlock patterns when capturing output.

## When to Use

- Running or orchestrating external processes in a .NET 11 (or later) project.

> - Needing to start a process, wait for it to exit, and capture its output/error streams without risking deadlocks (`Process.RunAndCaptureTextAsync`).

Reviewer: nit: we provide sync and async overloads

```suggestion
- Needing to start a process, wait for it to exit, and capture its output/error streams without risking deadlocks (`Process.RunAndCaptureText[Async]`).
```

- Wanting to ensure child processes are automatically terminated when the parent process exits (`KillOnParentExit`).

> - Requiring trimmer-friendly and NativeAOT-compatible process creation via `SafeProcessHandle`.

Reviewer: The wording needs a bit more specific. `Process` itself is NativeAOT-compatible, it's just that with `SafeProcessHandle` the user gets the smallest possible size on disk (so it's more of an optimization)

- Requiring fine-grained control over handle inheritance (`InheritedHandles`) or starting detached processes (`StartDetached`).

## When Not to Use

- The project targets .NET 10 or earlier — these APIs are not available before .NET 11.

> - Running simple shells where custom execution code is unnecessary.

Reviewer: I am not a native speaker, so please take it with a grain of salt. But overall I do believe that the new APIs simplify the code a lot, so I would recommend them even for simple shells.

- The default `Process.Start()` is sufficient and does not require output capturing or advanced lifecycle rules.

## Target Framework

```xml
<TargetFramework>net11.0</TargetFramework>
```

> ## New APIs & Convenience Methods

Reviewer: Each of these methods is a new API, some are also convenience methods. I would just call it `New APIs`

```suggestion
## New APIs
```

> ### High-Level Convenience APIs (Static Methods)

Reviewer: I would suggest we use following separation:

- New APIs
  - Convenience APIs
    - Static Methods
    - Instance Methods

As some of the instance methods are also convenience methods (at least in my opinion).

```suggestion
### High-Level Convenience APIs

#### Static Methods
```

#### `Process.Run` / `Process.RunAsync`

> Starts a process and waits for it to exit, returning the exit status. Does not capture standard output or error.

Reviewer: We need to mention that it also allows the users to discard the output/error by providing `silent: true` and internally redirecting std handles to `NUL` device.

public static Task<ProcessExitStatus> RunAsync(string fileName, IList<string>? arguments = null, bool silent = false, CancellationToken cancellationToken = default)
public static ProcessExitStatus Run(ProcessStartInfo startInfo, TimeSpan? timeout = null)
public static Task<ProcessExitStatus> RunAsync(ProcessStartInfo startInfo, CancellationToken cancellationToken = default)
```

#### `Process.RunAndCaptureText` / `Process.RunAndCaptureTextAsync`
Starts a process, captures both standard output and error, and waits for it to exit. Extremely useful for avoiding deadlocks on stream redirection.
```csharp
public static ProcessTextOutput RunAndCaptureText(string fileName, IList<string>? arguments = null, TimeSpan? timeout = null)
public static Task<ProcessTextOutput> RunAndCaptureTextAsync(string fileName, IList<string>? arguments = null, CancellationToken cancellationToken = default)
public static ProcessTextOutput RunAndCaptureText(ProcessStartInfo startInfo, TimeSpan? timeout = null)
public static Task<ProcessTextOutput> RunAndCaptureTextAsync(ProcessStartInfo startInfo, CancellationToken cancellationToken = default)
```

#### `Process.StartAndForget`

> Launches a process and immediately releases the system handle resources, returning only the process ID (PID).

Reviewer: Let's provide the reasoning behind this method. You can re-use the text from our [blog post](https://devblogs.microsoft.com/dotnet/process-api-improvements-in-dotnet-11/#process.startandforget):

```suggestion
There is a common misconception that when a process is disposed, it’s also being killed. This is not the case, as `Process.Dispose` only releases the resources associated with the process, but does not kill it.

To make it easier to start a process without the need to worry about disposing it, `Process.StartAndForget` was introduced. The method starts a process, returns its ID and releases all resources associated with it
```

```csharp

> public static int StartAndForget(string fileName, IList<string>? arguments = null)

Reviewer: In .NET 11 Preview 7 we have changed these methods to accept `IEnumerable<string>` for the sake of consistency with pre-existing Process APIs (https://github.com/dotnet/runtime/pull/130630)

```suggestion
public static ProcessExitStatus Run(string fileName, IEnumerable<string>? arguments = null, bool silent = false, TimeSpan? timeout = null)
public static Task<ProcessExitStatus> RunAsync(string fileName, IEnumerable<string>? arguments = null, bool silent = false, CancellationToken cancellationToken = default)
public static ProcessExitStatus Run(ProcessStartInfo startInfo, TimeSpan? timeout = null)
public static Task<ProcessExitStatus> RunAsync(ProcessStartInfo startInfo, CancellationToken cancellationToken = default)
```

#### `Process.RunAndCaptureText` / `Process.RunAndCaptureTextAsync`
Starts a process, captures both standard output and error, and waits for it to exit. Extremely useful for avoiding deadlocks on stream redirection.
```csharp
public static ProcessTextOutput RunAndCaptureText(string fileName, IEnumerable<string>? arguments = null, TimeSpan? timeout = null)
public static Task<ProcessTextOutput> RunAndCaptureTextAsync(string fileName, IEnumerable<string>? arguments = null, CancellationToken cancellationToken = default)
public static ProcessTextOutput RunAndCaptureText(ProcessStartInfo startInfo, TimeSpan? timeout = null)
public static Task<ProcessTextOutput> RunAndCaptureTextAsync(ProcessStartInfo startInfo, CancellationToken cancellationToken = default)
```

#### `Process.StartAndForget`
Launches a process and immediately releases the system handle resources, returning only the process ID (PID).
```csharp
public static int StartAndForget(string fileName, IEnumerable<string>? arguments = null)
```

public static int StartAndForget(ProcessStartInfo startInfo)
```

### Reliable Output Reading APIs (Instance Methods)
These methods are called on a `Process` instance to directly read stdout and stderr, guaranteeing no OS pipe buffer overflow deadlocks.

```csharp
public (string StandardOutput, string StandardError) ReadAllText(TimeSpan? timeout = null)
public Task<(string StandardOutput, string StandardError)> ReadAllTextAsync(CancellationToken cancellationToken = default)
public (byte[] StandardOutput, byte[] StandardError) ReadAllBytes(TimeSpan? timeout = null)
public Task<(byte[] StandardOutput, byte[] StandardError)> ReadAllBytesAsync(CancellationToken cancellationToken = default)
public IEnumerable<ProcessOutputLine> ReadAllLines(TimeSpan? timeout = null)
public IAsyncEnumerable<ProcessOutputLine> ReadAllLinesAsync(CancellationToken cancellationToken = default)
```

> *Note: `ProcessOutputLine` is a readonly struct containing `string Content` and `bool StandardError` properties.*

Reviewer: We mention what `ProcessOutputLine` is, but we don't mention `ProcessExitStatus` nor `ProcessTextOutput`. I think it's best to just provide their definition just before they appear for the first time.

### ProcessStartInfo Properties

#### `KillOnParentExit`

> Ensures that the spawned child process is terminated when the current (parent) process exits. Works across both Windows and Unix platforms.

Reviewer: Not every Unix is supported. Only Win, Lin and Android.

```suggestion
Ensures that the spawned child process is terminated when the current (parent) process exits. Works across Windows, Linux and Android.
```

```csharp
public bool KillOnParentExit { get; set; }
```

#### `InheritedHandles`

> Provides precise control over which file/kernel handles are inherited by the child process, preventing accidental resource leaks.

Reviewer: We need to describe few important things:
- std handles are always included (no need to add to the list)
- list set to empty list == only std handles get inerited
- only `SafeFileHandle` and `SafePipeHandle` instances are allowes as of today.
- no global lock used when spawning new process on Windows (important to tune projects that spawn multiple processes in parallel)

```csharp
public IList<SafeHandle>? InheritedHandles { get; set; }
```

#### `StartDetached`

> Starts the process detached from the parent's terminal or job session, ensuring it survives the parent's exit.

Reviewer: When set to `true`, the std handles are by default redirected to `NUL` device. So the child process does not keep the parent process console/terminal resources alive.

```csharp
public bool StartDetached { get; set; }
```

---

## Examples

### 1. One-Line Run and Capture Output

Run a process and safely read all output text without stream deadlock risks:

```csharp
using System;
using System.Diagnostics;
using System.Threading.Tasks;

// Run 'git status' and capture output (arguments passed as list)

> ProcessTextOutput result = await Process.RunAndCaptureTextAsync("git", ["status"]);

Reviewer: People tend to use `async` for apps that absolutely don't benefit from it (scalability, cancellation support etc). I think it's better to use non-async overload for such simple examples. So AI-written code is simple and does not pay for the price of using async
```suggestion
ProcessTextOutput result = Process.RunAndCaptureText("git", ["status"]);
```

if (result.ExitStatus.ExitCode == 0)
{

>     Console.WriteLine($"Git Output: {result.StandardOutput.Trim()}");

Reviewer: There is no need to use `Trim` here.

```suggestion
    Console.WriteLine($"Git Output: {result.StandardOutput}");
```

}
else
{
    Console.WriteLine($"Failed with exit code: {result.ExitStatus.ExitCode}");
    Console.WriteLine($"Error: {result.StandardError}");
}
```

### 2. Auto-Killing Child Processes on Parent Exit

Ensure a long-running background worker process is killed when the main application terminates:

```csharp
using System.Diagnostics;

> var startInfo = new ProcessStartInfo("dotnet", ["run", "--project", "BackgroundWorker.csproj"])

Reviewer: nit: subjective: no need to use `var`

```suggestion
ProcessStartInfo startInfo = new("dotnet", ["run", "--project", "BackgroundWorker.csproj"])
```

{

>     KillOnParentExit = true // Auto-teardown when this parent process exits

Reviewer: ```suggestion
    KillOnParentExit = OperatingSystem.IsWindows() ||  OperatingSystem.IsLinux() // Auto-teardown when this parent process exits
```
